internal/kafka/circuitbreaker.go

```diff
@@ -0,0 +1,184 @@
+}
+
+// circuitBreaker checks that the
+type circuitBreaker struct {
+	conf  *CircuitBreakerConf
+	mux   sync.Mutex
+	state map[string]map[int32]*cbPartitionState
+}
+
+var singletonCircuitBreaker *circuitBreaker
+
+func InitCircuitBreaker(conf *CircuitBreakerConf) error {
+
+	if conf == nil || !conf.Enabled {
+		return nil
+	}
+	if singletonCircuitBreaker != nil {
+		log.Warnf("Multiple Kafka Consumer CircuitBreaker configurations in config. Only the first will be applied")
+		return nil
+	}
+
+	cb := &circuitBreaker{
+		conf:  conf,
+		state: make(map[string]map[int32]*cbPartitionState),
+	}
+	cb.conf.logFrequency = time.Duration(cb.conf.LogFrequencySec) * time.Second
+	if cb.conf.logFrequency <= 0 {
+		cb.conf.logFrequency = DefaultLogFrequency
+	}
+	if cb.conf.UpperBound <= 0 {
+		cb.conf.UpperBound = DefaultUpperBound
+	}
+	if cb.conf.ResetThreshold <= 0 || cb.conf.ResetThreshold > 1 {
+		cb.conf.ResetThreshold = DefaultResetThreshold
+	}
+	cb.conf.resetBufferSize = int64(math.Ceil(cb.conf.ResetThreshold * float64(cb.conf.UpperBound)))
+	singletonCircuitBreaker = cb
+	log.Infof("CircuitBreakerEnabled: trip=%.2fKb reset=%.2fKb", float64(cb.conf.UpperBound)/1024, float64(cb.conf.resetBufferSize)/1024)
+
+	return nil
+}
+
+func GetCircuitBreaker() *circuitBreaker {
+	return singletonCircuitBreaker
+}
+
+func (cb *circuitBreaker) logState(prefix, topic string, partition int32, partitionState *cbPartitionState) {
+	lastTripped := partitionState.lastTripTime.Format(time.RFC3339Nano)
+	if partitionState.lastTripTime.IsZero() {
+		lastTripped = "never"
+	}
+	log.Infof("%s: topic=%s partition=%d offset=%d hwm=%d gap=%d gap.estimate=%.2fKb tripped=%t lastTripped=%s",
+		prefix, topic, partition, partitionState.offset, partitionState.hwm, partitionState.gap, float64(partitionState.bufSize)/1024, partitionState.tripped, lastTripped)
+}
+
+func (cb *circuitBreaker) Update(topic string, partition int32, hwm, offset, size int64) {
+	cb.mux.Lock()
+	defer cb.mux.Unlock()
+
+	topicState := cb.state[topic]
+	if topicState == nil {
+		topicState = make(map[int32]*cbPartitionState)
+		cb.state[topic] = topicState
+	}
+	partitionState := topicState[partition]
+	if partitionState == nil {
+		partitionState = &cbPartitionState{
+			msgCount:     1,
+			msgBytes:     size,
+			sizeEstimate: size,
+		}
+		topicState[partition] = partitionState
+	} else {
+		partitionState.msgCount++
+		partitionState.msgBytes += size
+		partitionState.sizeEstimate = partitionState.msgBytes / partitionState.msgCount
+	}
+	partitionState.hwm = hwm
+	partitionState.offset = offset
+	partitionState.gap = (hwm - offset)
+	if partitionState.gap <= 0 {
+		partitionState.gap = 0 // ensure positive
+	}
+	partitionState.bufSize = (partitionState.gap * partitionState.sizeEstimate)
+
+	if partitionState.tripped {
+		if partitionState.bufSize < cb.conf.resetBufferSize {
+			partitionState.tripped = false
+			cb.logState("CircuitBreakerReset", topic, partition, partitionState)
```

Comment:
Wondering if a `lastResetTime` would be useful as well in diagnosing problems with chatty/bursty submissions.